pkg/sloop/ingress/kubewatcher.go

```diff
@@ -306,14 +312,32 @@ func (i *kubeWatcherImpl) processUpdate(kind string, obj interface{}, watchResul
 	glog.V(99).Infof("processUpdate: obj json: %v", resourceJson)
 
 	kubeMetadata, err := kubeextractor.ExtractMetadata(resourceJson)
+	EventInfo, err1 := kubeextractor.ExtractEventInfo(resourceJson)
+	InvolvedObject, err2 := kubeextractor.ExtractInvolvedObject(resourceJson)
 	if err != nil || kubeMetadata.Namespace == "" {
 		// We are only grabbing namespace here for a prometheus metric, so if metadata extract fails we just log and continue
 		glog.V(2).Infof("No namespace for resource: %v", err)
 	}
-	metricIngressKubewatchcount.WithLabelValues(kind, watchResult.WatchType.String(), kubeMetadata.Namespace).Inc()
-	metricIngressKubewatchbytes.WithLabelValues(kind, watchResult.WatchType.String(), kubeMetadata.Namespace).Add(float64(len(resourceJson)))
+	if err1 != nil {
+		glog.V(2).Infof("Extract event info: %v", err1)
+	}
+	if err2 != nil {
+		glog.V(2).Infof("Error occured while extracting Involved Object Info: %v", err2)
+	}
+	if enableGranularmetrics {
+
+		if kind == "Event" {
+			metricIngressGranularKubewatchcount.WithLabelValues(kind, watchResult.WatchType.String(), kubeMetadata.Namespace, kubeMetadata.Name, InvolvedObject.Kind, EventInfo.Reason, EventInfo.Type).Inc()
+			metricIngressGranularKubewatchbytes.WithLabelValues(kind, watchResult.WatchType.String(), kubeMetadata.Namespace, kubeMetadata.Name, InvolvedObject.Kind, EventInfo.Reason, EventInfo.Type).Add(float64(len(resourceJson)))
+		}
+		metricIngressKubewatchcount.WithLabelValues(kind, watchResult.WatchType.String(), kubeMetadata.Namespace).Inc()
+		metricIngressKubewatchbytes.WithLabelValues(kind, watchResult.WatchType.String(), kubeMetadata.Namespace).Add(float64(len(resourceJson)))
+	} else {
+		metricIngressKubewatchcount.WithLabelValues(kind, watchResult.WatchType.String(), kubeMetadata.Namespace).Inc()
+		metricIngressKubewatchbytes.WithLabelValues(kind, watchResult.WatchType.String(), kubeMetadata.Namespace).Add(float64(len(resourceJson)))
+	}
 
-	glog.V(common.GlogVerbose).Infof("Informer update (%s) - Name: %s, Namespace: %s, ResourceVersion: %s", watchResult.WatchType, kubeMetadata.Name, kubeMetadata.Namespace, kubeMetadata.ResourceVersion)
+	glog.V(common.GlogVerbose).Infof("Informer update (%s) - Name: %s, Namespace: %s, ResourceVersion: %s, Reason: %s, Type: %s", watchResult.WatchType, kubeMetadata.Name, kubeMetadata.Namespace, kubeMetadata.ResourceVersion, EventInfo.Reason, EventInfo.Type)
```

Comment:
Can a nil value of EventInfo cause any issue in this logline for EventInfo.Reason & Type or is it handled implicitly in this glog.V?

Reply:
@anuj-anurag-sharma I do not think extractEvent Info can be nil , looking at the events.go and events_test.go looks like it ignores unavailable fields and keeps it empty string. Hence those values would be empty string but not nil.

Reply:
And there are fields like timestamps which even if not mentioned in payload json are automatically generated as time.Time() in the extractevent Info function